### PR TITLE
Mecha refactor 2020

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

image

## Changes

-Mechs are a vehicle subtype
-Mech equipment half-rewritten
-Mech actions completely redone
-Cooldown macros
-New movement macros & replacing all var in GLOB.diagonals with ISDIAGONALDIR(var)
-New lazylist macro
-Support for lavaland only mechs
-Removed the tank because fuck off with that hacky shit
-Documentation
-Fuckton of fixes
-More shit i forgot

##ToDo
opening as a draft so I can get early reviews in
- [x] Bitflags?
- [x] Test and doubletest
- [x]  Change more hacky overrides to signals

## Changelog
:cl:
refactor: Mech refactor
/:cl:

### CODE REVIEW

**Comment by TheSilverNuke** [comment]
I always appreciate mechs getting more attention.

**Comment by ATH1909** [comment]
holy shit he's touching mech code

godspeed, sir. we'll pray for your sanity.

**Comment by necromanceranne** [comment]
Goodness gracious. Everything is changing!!!

**Comment by OmegaDarkPotato** [comment]
Hopefully we get some really busted lavaland only mechs with this - Fighting megafauna specifically designed for mech vs monster combat would be pretty neato.

**Comment by LemonInTheDark** [comment]
You absolute madman

**Comment by TheSilverNuke** [comment]
> Fighting megafauna specifically designed for mech vs monster combat would be pretty neato.

Pacific Rim SS13 Edition

File: code/__DEFINES/vehicles.dm [hunk 1]
```diff
@@ -7,3 +7,10 @@
 
 //Car trait flags
 #define CAN_KIDNAP 1
+
+#define ISDIAGONALDIR(d) (d&(d-1))
+#define NSCOMPONENT(d)   (d&(NORTH|SOUTH))
+#define EWCOMPONENT(d)   (d&(EAST|WEST))
+#define NSDIRFLIP(d)     (d^(NORTH|SOUTH))
+#define EWDIRFLIP(d)     (d^(EAST|WEST))
+#define DIRFLIP(d)       turn(d, 180)
```
Comment (Rohesie) [inline code/__DEFINES/vehicles.dm:16]: Put this in `flags.dm`

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -359,7 +359,7 @@ Turf and target are separate in case you want to teleport some distance from a t
 		x = world.maxx
 	else if(direction & WEST)
 		x = 1
-	if(direction in GLOB.diagonals) //let's make sure it's accurately-placed for diagonals
+	if(ISDIAGONALDIR(direction)) //let's make sure it's accurately-placed for diagonals
```
Comment (Rohesie) [inline code/__HELPERS/unsorted.dm:362]: If you could atomize this change, we'd be able to merge it rather fast and it'd diminish the number of changed files here.

**Comment by Rohesie** [comment]
You'll have to poke the keyholders to get it actually testmerged, and to get feedback from that. I'll slap the label as needed. Feel free to poke me about it.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 3]
```diff
@@ -215,10 +215,11 @@
...
 		if(ismecha(the_target))
-			var/obj/mecha/M = the_target
-			if(M.occupant)//Just so we don't attack empty mechs
-				if(CanAttack(M.occupant))
-					return TRUE
+			var/obj/vehicle/sealed/mecha/M = the_target
+			if(LAZYLEN(M.occupants))//Just so we don't attack empty mechs
+				for(var/occupant in M.occupants)
+					if(CanAttack(occupant))
+						return TRUE
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:222]:
```suggestion
			for(var/occupant in M.occupants)
				if(CanAttack(occupant))
					return TRUE
```
`for(var/i in null)` will simply not execute, as expected

**Comment by optimumtact** [comment]
this is exceptionally well done btw.

**Comment by carshalash** [comment]
Noticing on live servers that mechs are now able to shoot twice as much/as fast as they were before.

**Comment by Iamgoofball** [comment]
just found out this guy removed the tank, this means war

**Pull request merged**